File: MasteringDart/lib/chapters/chapter12/advanced/advanced_error_handling.dart
// ignore_for_file: avoid_print
//
//  advanced_error_handling.dart
//  MasteringDart
//
//  Capítulo 12 — Manejo de errores.
//  Nivel AVANZADO: Validación UserRegistration — username, email, password.
//
//  Ejecutar con: dart run lib/chapters/chapter12/advanced/advanced_error_handling.dart

/// Excepción base para errores de registro.
class RegistrationException implements Exception {
  final String message;
  RegistrationException(this.message);
  @override
  String toString() => message;
}

class InvalidUsernameException extends RegistrationException {
  InvalidUsernameException(super.message);
}

class InvalidEmailException extends RegistrationException {
  InvalidEmailException(super.message);
}

class InvalidPasswordException extends RegistrationException {
  InvalidPasswordException(super.message);
}

class UserRegistration {
  static void validateUsername(String username) {
    if (username.length < 3) {
      throw InvalidUsernameException(
        "Username must be at least 3 characters: '$username'",
      );
    }
  }

  static void validateEmail(String email) {
    if (!email.contains('@') || !email.contains('.')) {
      throw InvalidEmailException("Invalid email format: '$email'");
    }
  }

  static void validatePassword(String password) {
    if (password.length < 6) {
      throw InvalidPasswordException('Password must be at least 6 characters');
    }
  }

  static void register(String username, String email, String password) {
    validateUsername(username);
    validateEmail(email);
    validatePassword(password);
    print('Registration successful: $username');
  }
}

void runAdvancedErrorHandlingExamples() {
  print('========================================');
  print('       USER REGISTRATION VALIDATION');
  print('========================================');

  // --- Registro válido ---
  try {
    UserRegistration.register('alice', 'alice@mail.com', 'secret123');
  } on RegistrationException catch (e) {
    print('Error: $e');
  }

  // --- Username inválido ---
  try {
    UserRegistration.register('ab', 'bob@mail.com', 'pass123');
  } on InvalidUsernameException catch (e) {
    print('Username error: $e');
  } on RegistrationException catch (e) {
    print('Error: $e');
  }

  // --- Email inválido ---
  try {
    UserRegistration.register('charlie', 'invalid-email', 'pass123');
  } on InvalidEmailException catch (e) {
    print('Email error: $e');
  } on RegistrationException catch (e) {
    print('Error: $e');
  }

  // --- Password inválido ---
  try {
    UserRegistration.register('diana', 'diana@mail.com', '123');
  } on InvalidPasswordException catch (e) {
    print('Password error: $e');
  } on RegistrationException catch (e) {
    print('Error: $e');
  }

  print('========================================');
}

void main() {
  runAdvancedErrorHandlingExamples();
}

File: MasteringKotlin/app/src/main/java/com/devlokos/masteringkotlin/chapters/chapter12/advanced/AdvancedErrorHandling.kt
package com.devlokos.masteringkotlin.chapters.chapter12.advanced

/**
 * Capítulo 12 — Manejo de Errores.
 * Nivel AVANZADO: Validación UserRegistration — username, email, password; throw y manejar de forma segura.
 */
fun main() {
    runAdvancedErrorHandlingExamples()
}

/** Excepción base para errores de registro. */
open class RegistrationException(message: String) : Exception(message)

/** Username demasiado corto o inválido. */
class InvalidUsernameException(message: String) : RegistrationException(message)

/** Formato de email inválido. */
class InvalidEmailException(message: String) : RegistrationException(message)

/** Contraseña demasiado débil. */
class InvalidPasswordException(message: String) : RegistrationException(message)

object UserRegistration {
    fun validateUsername(username: String) {
        if (username.length < 3) {
            throw InvalidUsernameException("Username must be at least 3 characters: '$username'")
        }
    }

    fun validateEmail(email: String) {
        if (!email.contains("@") || !email.contains(".")) {
            throw InvalidEmailException("Invalid email format: '$email'")
        }
    }

    fun validatePassword(password: String) {
        if (password.length < 6) {
            throw InvalidPasswordException("Password must be at least 6 characters")
        }
    }

    fun register(username: String, email: String, password: String) {
        validateUsername(username)
        validateEmail(email)
        validatePassword(password)
        println("Registration successful: $username")
    }
}

fun runAdvancedErrorHandlingExamples() {
    println("========================================")
    println("       USER REGISTRATION VALIDATION")
    println("========================================")

    // --- Registro válido ---
    try {
        UserRegistration.register("alice", "alice@mail.com", "secret123")
    } catch (e: RegistrationException) {
        println("Error: ${e.message}")
    }

    // --- Username inválido ---
    try {
        UserRegistration.register("ab", "bob@mail.com", "pass123")
    } catch (e: InvalidUsernameException) {
        println("Username error: ${e.message}")
    } catch (e: RegistrationException) {
        println("Error: ${e.message}")
    }

    // --- Email inválido ---
    try {
        UserRegistration.register("charlie", "invalid-email", "pass123")
    } catch (e: InvalidEmailException) {
        println("Email error: ${e.message}")
    } catch (e: RegistrationException) {
        println("Error: ${e.message}")
    }

    // --- Password inválido ---
    try {
        UserRegistration.register("diana", "diana@mail.com", "123")
    } catch (e: InvalidPasswordException) {
        println("Password error: ${e.message}")
    } catch (e: RegistrationException) {
        println("Error: ${e.message}")
    }

    println("========================================")
}
